### PR TITLE
cmd/clone:  clean the detached nodes in the gc command 

### PR DESCRIPTION
**Pull request by zhijian-pro**

### CODE REVIEW

File: pkg/meta/base.go [hunk 1]
```diff
@@ -1977,3 +1994,58 @@ func (m *baseMeta) ScanDeletedObject(ctx Context, tss trashSliceScan, pss pendin
 	}
 	return eg.Wait()
 }
+
+func (m *baseMeta) Clone(ctx Context, srcIno, dstParentIno Ino, dstName string, cmode uint8, cumask uint16, count, total *uint64) syscall.Errno {
```
Comment (davies) [inline pkg/meta/base.go:1998]: Can we move these refactor to another PR?

Comment (zhijian-pro) [reply pkg/meta/base.go]: refactor depends on this PR, which will come out after this merger.

File: pkg/meta/redis.go [hunk 2]
```diff
@@ -4311,3 +4301,37 @@ func (m *redisMeta) mkNodeWithAttr(ctx Context, tx *redis.Tx, srcIno Ino, srcAtt
...
+	exists, err := m.rdb.Exists(ctx, m.inodeKey(detachedNode)).Result()
+	if err != nil {
+		return errno(err)
+	}
+	if exists == 1 {
+		rmConcurrent := make(chan int, 10)
+		if eno := m.emptyDir(ctx, detachedNode, true, nil, rmConcurrent); eno != 0 {
+			return eno
+		}
+		if err := m.txn(ctx, func(tx *redis.Tx) error {
```
Comment (davies) [inline pkg/meta/redis.go:4315]: the node could be a file

File: pkg/meta/sql.go [hunk 3]
```diff
@@ -166,6 +166,11 @@ type dirStats struct {
 	UsedInodes int64 `xorm:"notnull"`
 }
 
+type detachedNode struct {
+	Inode  Ino   `xorm:"pk notnull"`
+	Expire int64 `xorm:"notnull"`
```
Comment (davies) [inline pkg/meta/sql.go:171]: Added

File: pkg/meta/tkv.go [hunk 4]
```diff
@@ -184,6 +184,7 @@ All keys:
   SHssssssss         session heartbeat // for legacy client
   SIssssssss         session info
   SSssssssssiiiiiiii sustained inode
+  Niiiiiiii	         detached inodes
```
Comment (davies) [inline pkg/meta/tkv.go:187]: align

File: pkg/meta/redis.go (hunk 2 above)
Comment (zhijian-pro) [reply pkg/meta/redis.go:4315]: The detached node  is logged into the metadata engine only when `attach = false`.
`attach` is always true if the clone object is a file.
```
mkNodeWithAttr
...
if !attach {
    pipe.ZAdd(ctx, m.detachedNodes(), redis.Z{Member: dstIno.String(), Score: float64(time.Now().Unix())})
}
...
```

File: pkg/meta/tkv.go (hunk 4 above)
Comment (zhijian-pro) [reply pkg/meta/tkv.go]: image
image

github display cannot be aligned with the local IDE at the same time

Comment (zhijian-pro) [reply pkg/meta/tkv.go]: vim
image
I think the local ide presentation is right.

Comment (zhijian-pro) [reply pkg/meta/tkv.go]: Maybe it was a special character problem before, so delete it and rewrite it

File: pkg/meta/sql.go [hunk 5]
```diff
@@ -4007,9 +3996,56 @@ func (m *dbMeta) mkNodeWithAttr(ctx Context, s *xorm.Session, srcIno Ino, srcNod
...
+			}
+			if _, err := s.Delete(&xattr{Inode: detachedNode}); err != nil {
+				return err
+			}
+			return nil
+		}); err != nil {
+			return errno(err)
+		}
+	}
+	_, err = m.db.Delete(&node{Inode: detachedNode})
```
Comment (davies) [inline pkg/meta/sql.go:4049]: when it exists

File: pkg/meta/sql.go [hunk 6]
```diff
@@ -4007,9 +3996,56 @@ func (m *dbMeta) mkNodeWithAttr(ctx Context, s *xorm.Session, srcIno Ino, srcNod
 		}
 	}
 
+	if !attach {
+		return mustInsert(s, &detachedNode{Inode: *dstIno, Added: time.Now().Unix()})
```
Comment (davies) [inline pkg/meta/sql.go:4000]: move to the end

File: pkg/meta/redis.go [hunk 7]
```diff
@@ -4299,6 +4286,9 @@ func (m *redisMeta) mkNodeWithAttr(ctx Context, tx *redis.Tx, srcIno Ino, srcAtt
 
 	_, err = tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
 		pipe.Set(ctx, m.inodeKey(*dstIno), m.marshal(srcAttr), 0)
+		if !attach {
+			pipe.ZAdd(ctx, m.detachedNodes(), redis.Z{Member: dstIno.String(), Score: float64(time.Now().Unix())})
```
Comment (davies) [inline pkg/meta/redis.go:4290]: move to the end
```
if attach {
} else {
}
```

File: pkg/meta/tkv.go [hunk 8]
```diff
@@ -3470,6 +3458,7 @@ func (m *kvMeta) mkNodeWithAttr(ctx Context, tx *kvTxn, srcIno Ino, srcAttr *Att
 		}
 	}
 
+	tx.set(m.detachedKey(*dstIno), m.packInt64(time.Now().Unix()))
```
Comment (davies) [inline pkg/meta/tkv.go:3461]: when !attach

File: pkg/meta/base.go [hunk 9]
```diff
@@ -70,6 +70,8 @@ type engine interface {
 	doCleanupSlices()
 	doCleanupDelayedSlices(edge int64) (int, error)
 	doDeleteSlice(id uint64, size uint32) error
+	doFindDetachedNodes(t time.Time) []Ino
```
Comment (davies) [inline pkg/meta/base.go:73]: add unit test for these

File: pkg/meta/sql.go (hunk 5 above)
Comment (davies) [reply pkg/meta/sql.go:4049]: you don't need to delete it, if not existed

File: pkg/meta/redis.go [hunk 10]
```diff
@@ -4306,8 +4296,44 @@ func (m *redisMeta) mkNodeWithAttr(ctx Context, tx *redis.Tx, srcIno Ino, srcAtt
 			tx.HSet(ctx, m.entryKey(dstParentIno), dstName, m.packEntry(srcAttr.Typ, *dstIno))
 			tx.IncrBy(ctx, m.usedSpaceKey(), align4K(0))
 			tx.Incr(ctx, m.totalInodesKey())
+		} else {
+			pipe.ZAdd(ctx, m.detachedNodes(), redis.Z{Member: dstIno.String(), Score: float64(time.Now().Unix())})
```
Comment (zhijian-pro) [inline pkg/meta/redis.go:4300]: This action needs to be done in a transaction and cannot be presented as a common method.
